liteeth/mac/sram.py

```diff
@@ -62,23 +68,58 @@ def __init__(self, dw, depth, nslots=2, endianness="big"):
         ongoing = Signal()
 
         # Status FIFO
-        fifo = stream.SyncFIFO([("slot", slotbits), ("length", lengthbits)], nslots)
+        if timestamp_source is not None:
+            fifo = stream.SyncFIFO([
+                ("slot", slotbits),
+                ("length", lengthbits),
+                ("rx_timestamp", timestampbits),
+            ], nslots)
+        else:
+            fifo = stream.SyncFIFO([("slot", slotbits), ("length", lengthbits)], nslots)
         self.submodules += fifo
 
+        # If we timestamp incoming packets we're going to need another
+        # signal to hold the timestamp from the start of packet
+        # reception
+        if timestamp_source is not None:
+            rx_start_timestamp = Signal(timestampbits)
+
+            # Effectively build a latch, updating the timestamp only
+            # when starting the reception
+            self.comb += [
+                rx_start_timestamp.eq(rx_start_timestamp)
+            ]
+
         # FSM
         self.submodules.fsm = fsm = FSM(reset_state="IDLE")
-        fsm.act("IDLE",
-            If(sink.valid,
-                If(fifo.sink.ready,
-                    ongoing.eq(1),
-                    NextValue(counter, counter + inc),
-                    NextState("WRITE")
-                ).Else(
-                    NextValue(self.errors.status, self.errors.status + 1),
-                    NextState("DISCARD_REMAINING")
+        if timestamp_source is not None:
+            fsm.act("IDLE",
+                If(sink.valid,
+                    If(fifo.sink.ready,
+                        ongoing.eq(1),
+                        rx_start_timestamp.eq(timestamp_source),
+                        NextValue(counter, counter + inc),
+                        NextState("WRITE")
+                    ).Else(
+                        NextValue(self.errors.status, self.errors.status + 1),
+                        NextState("DISCARD_REMAINING")
+                    )
                 )
             )
-        )
+        else:
+            fsm.act("IDLE",
+                If(sink.valid,
+                    If(fifo.sink.ready,
+                        ongoing.eq(1),
+                        NextValue(counter, counter + inc),
+                        NextState("WRITE")
+                    ).Else(
+                        NextValue(self.errors.status, self.errors.status + 1),
+                        NextState("DISCARD_REMAINING")
+                    )
+                )
+            )
+
```

Comment:
We should avoid duplicating too much FSM code, this can be simplified with something like:
````python3          
    fsm.act("IDLE",
                If(sink.valid,
                    If(fifo.sink.ready,
                        ongoing.eq(1),
                        *[] if timestamp_source is None else *[rx_start_timestamp.eq(timestamp_source)],
                        NextValue(counter, counter + inc),
                        NextState("WRITE")
                    ).Else(
                        NextValue(self.errors.status, self.errors.status + 1),
                        NextState("DISCARD_REMAINING")
                    )
                )
            )
````